crypto/pqt/pqt_kem.c

```diff
@@ -0,0 +1,512 @@
+
+// NIST-P secret keys are scalars. This function writes the secret key to
+// |secret_key| as a big-endian integer, padded with zeros to length
+// |secret_key_len|.
+//
+// It matches SerializePrivateKey in RFC 9180.
+static int nistp_serialize_secret_key(uint8_t *secret_key,
+                                      size_t secret_key_len,
+                                      const EC_KEY *eckey) {
+  return BN_bn2bin_padded(secret_key, secret_key_len,
+                          EC_KEY_get0_private_key(eckey));
+}
+
+// This function parses the |secret_key| buffer back into a scalar, checks
+// that it is not zero, and returns a freshly allocated |EC_KEY| on success,
+// and NULL on error.
+//
+// It matches DeserializePrivateKey in RFC 9180.
+static EC_KEY *nistp_deserialize_secret_key(const uint8_t *secret_key,
+                                            size_t secret_key_len,
+                                            const EC_GROUP *group) {
+  BIGNUM *secret_key_num = BN_bin2bn(secret_key, secret_key_len, NULL);
+  EC_KEY *eckey = EC_KEY_new();
+  if ((secret_key_num == NULL) || (eckey == NULL) ||
+      !EC_KEY_set_group(eckey, group) ||
+      !EC_KEY_set_private_key(eckey, secret_key_num)) {
+    BN_free(secret_key_num);
+    EC_KEY_free(eckey);
+    return NULL;
+  }
+
+  BN_free(secret_key_num);
+  return eckey;
+}
+
+// NIST-P public keys are elliptic curve points. This function writes the public
+// key to |public_key| as an uncompressed point, to length |secret_key_len|.
+//
+// It matches SerializePublicKey in RFC 9180.
+static int nistp_serialize_public_key(uint8_t *public_key,
+                                      size_t public_key_len,
+                                      const EC_KEY *eckey) {
+  return (EC_POINT_point2oct(EC_KEY_get0_group(eckey),
+                             EC_KEY_get0_public_key(eckey),
+                             POINT_CONVERSION_UNCOMPRESSED, public_key,
+                             public_key_len, NULL) == public_key_len);
+}
+
+// This function parses the |public_key| buffer back into an elliptic curve
+// point, and returns a freshly allocated |EC_KEY| on success, and NULL on
+// error. It does not perform validity checks on the public key.
+//
+// Its API matches DeserializePrivateKey in RFC 9180.
+static EC_KEY *nistp_deserialize_public_key_unchecked(const uint8_t *public_key,
+                                                      size_t public_key_len,
+                                                      const EC_GROUP *group) {
+  EC_POINT *point = EC_POINT_new(group);
+  EC_KEY *eckey = EC_KEY_new();
+  if ((point == NULL) || (eckey == NULL) ||
+      !EC_POINT_oct2point(group, point, public_key, public_key_len, NULL) ||
+      !EC_KEY_set_group(eckey, group) || !EC_KEY_set_public_key(eckey, point)) {
+    EC_POINT_free(point);
+    EC_KEY_free(eckey);
+    return NULL;
+  }
+  EC_POINT_free(point);
+  return eckey;
+}
+
+static int nistp_keygen_deterministic(const EC_GROUP *group,
```

Comment:
can you please add a reference to the definition of keygen/encaps/decaps as implemented below?

Reply:
Added docstrings in the latest version.